src/ic_layer.cpp

```diff
@@ -0,0 +1,41 @@
+#include <ic_layer.h>
+
+namespace IC {
+    Layer::Layer(const std::string &debugName) : _debugName(debugName) {}
+
+    Layer::~Layer() {}
+
+    LayerStack::LayerStack() {
+        _layerInsertIndex = _layers.begin();
+    }
+
+    LayerStack::~LayerStack() {
+        for (Layer *layer : _layers)
+            delete layer;
+    }
+
+    void LayerStack::PushLayer(Layer *layer) {
+        _layerInsertIndex = _layers.emplace(_layerInsertIndex, layer);
+    }
+
+    void LayerStack::PushOverlay(Layer *overlay) {
+        _layers.emplace_back(overlay);
+    }
+
+    void LayerStack::PopLayer(Layer *layer) {
```

Comment:
Something to think about with these `Pop` functions: When we push a Layer or Overlay onto the LayerStack we expecting the Layer to be allocated for us, and we are taking direct control over this memory. Currently, if the LayerStack gets destructed we will properly clean up this memory, but if the Layer is popped we are giving control of this memory to the calling function.

Reply:
As mentioned; Ok with the assumption that right now if you pop a layer you'll have to de-allocate the memory after the fact.